src/MongoDB.Driver.Core/Core/Servers/TopologyDescription.cs

```diff
@@ -0,0 +1,323 @@
+        {
+            _processId = (ObjectId) info.GetValue("_processId", typeof(ObjectId));
+            _counter = (long) info.GetValue("_counter", typeof(long));
+            _hashCode = new Hasher().Hash(_processId).Hash(_counter).GetHashCode();
+        }
+#endif
+
+        // properties
+        /// <summary>
+        /// Gets the process identifier.
+        /// </summary>
+        /// <value>
+        /// The process identifier.
+        /// </value>
+        public ObjectId ProcessId => _processId;
+
+        /// <summary>
+        /// Gets the counter.
+        /// </summary>
+        /// <value>
+        /// The counter.
+        /// </value>
+        public long Counter => _counter;
+
+        // methods
+        /// <summary>
+        /// Gets whether <paramref name="x"/>.Equals(<paramref name="y"/>).
+        /// </summary>
+        /// <param name="x">A TopologyDescription.</param>
+        /// <param name="y">A TopologyDescription.</param>
+        /// <returns>
+        /// Whether <paramref name="x"/>.Equals(<paramref name="y"/>).
+        /// </returns>
+        public static bool operator ==(TopologyDescription x, TopologyDescription y) => CompareFreshnessOfLocalToServerResponse(x, y) == 0;
+
+        /// <summary>
+        /// Gets whether <paramref name="x"/>  != (<paramref name="y"/>).
+        /// </summary>
+        /// <param name="x">A TopologyDescription.</param>
+        /// <param name="y">A TopologyDescription.</param>
+        /// <returns>
+        /// Whether <paramref name="x"/> != (<paramref name="y"/>).
+        /// </returns>
+        public static bool operator !=(TopologyDescription x, TopologyDescription y) => CompareFreshnessOfLocalToServerResponse(x, y) != 0;
+
+        /// <summary>
+        /// Compares a local TopologyDescription with aserver's TopologyDescription and indicates whether the local
+        /// TopologyDescription is staler, fresher, or equal to the server's TopologyDescription.
+        /// Per the SDAM specification, if the ProcessIds are not equal, this method assumes that
+        /// <paramref name="serverResponse"/> is more recent. This means that this method does not exhibit
+        /// the reversal properties of inequalities i.e. a "&lt;" b does not imply b "&gt;" a.
+        /// </summary>
+        /// <param name="local"> The locally saved TopologyDescription. </param>
+        /// <param name="serverResponse">The TopologyDescription received from a server.</param>
+        /// <returns>
+        /// Less than zero indicates that the local description is older than the response.
+        /// Zero indicates that the local description is equal to the response.
+        /// Greater than zero indicates that the local description is newer than the response.
+        /// </returns>
+        public static int CompareFreshnessOfLocalToServerResponse(TopologyDescription local, TopologyDescription serverResponse)
```

Comment:
So, IMHO, comparing TopologyDescriptions is a very error prone process as much of "intuition" here for CompareTo or the inequality operators doesn't apply particularly well. That being said, I've tried to simplify the logic by splitting it up into two parts. The first part, here, compares `TopologyDescription`s, which because they are structs, cannot be null. The second part, in the extension methods below, compares `Nullable<TopologyDescription>`s, which CAN be null and has additional logic. It may be helpful to think of `TopologyDescription` as a base class and Nullable<TopologyDescription> as subclass with overriden "CompareTo" logic.